hathor/p2p/sync_v2/agent.py

```diff
@@ -545,36 +549,42 @@ def find_best_common_block(self,
         # Run an n-ary search in the interval [lo, hi).
         # `lo` is always a height where we are synced.
         # `hi` is always a height where sync state is unknown.
-        hi = min(peer_best_block.height, my_best_block.height)
-        lo = 0
-
-        lo_block_hash = self._settings.GENESIS_BLOCK_HASH
+        hi = min(peer_best_block, my_best_block)
+        lo = _HeightInfo(height=0, id=self._settings.GENESIS_BLOCK_HASH)
 
-        while hi - lo > 1:
+        while hi.height - lo.height > 1:
             self.log.info('find_best_common_block n-ary search query', lo=lo, hi=hi)
-            step = math.ceil((hi - lo) / 10)
-            heights = list(range(lo, hi, step))
-            heights.append(hi)
-
-            block_height_list = yield self.get_peer_block_hashes(heights)
-            block_height_list.sort(key=lambda x: x.height, reverse=True)
-
-            for height, block_hash in block_height_list:
+            step = math.ceil((hi.height - lo.height) / 10)
+            heights = list(range(lo.height, hi.height, step))
+            heights.append(hi.height)
+
+            block_info_list = yield self.get_peer_block_hashes(heights)
+            block_info_list.sort(key=lambda x: x.height, reverse=True)
+
+            # As we are supposed to be always synced at `lo`, we expect to receive a response
+            # with at least one item equals to lo. If it does not happen, we stop the search
+            # and return None. This might be caused when a reorg occurs during the search.
+            if not block_info_list:
+                return None
+            lo_block_info = block_info_list[-1]
+            if lo_block_info != lo:
+                return None
```

Comment:
**suggestion:** Do not use `None` to convey meaning.

I think those are clear and specific cases that can happen, and they have a known reason. We should only use `None` for the absence of value, but here it's representing information. We also loose information on which case happened, as they both return `None` (it isn't clear to me if the second case is also caused by a reorg).

In my opinion it would be better for future debugging and maintainability to create specific types that represent the information that those failures happened. They could be as simple as an empty NamedTuple/dataclass (these are known as zero-sized types). Then the return type would be something like `_HeightInfo | ReorgFailure | SomeOtherError`. That value can be easily matched against later, if needed.

Also, maybe add some logs?

Reply:
I really like typed errors on the return values. I'd prefer to introduce them using some ergonomic lib though, to make use of match. To me it's enough if we log the "reason" even if it still returns `None`.

Reply:
I added log messages for these cases. Another possibility is to raise custom exceptions instead of returning None. What do you think?

Reply:
Because the use of this method is very local, I think an exception would just make it more annoying to use. But also it's not too bad.

Reply:
> I really like typed errors on the return values. I'd prefer to introduce them using some ergonomic lib though, to make use of match.

@jansegre In this case it's not necessarily an error, but a possible state that is considered invalid. We don't really need a lib for some simple use cases, for example:

```python
@dataclass
class NoDataState:
    pass

@dataclass
class SomeIntState:
    a: int

@dataclass
class SomeStrState:
    b: str

MyState: TypeAlias = NoDataState | SomeIntState | SomeStrState

def handle_state(state: MyState) -> str:
    match state:
        case NoDataState(): return "no data in this state."
        case SomeIntState(some_int): return f"state with number: {some_int}"
        case SomeStrState(some_str): return some_str
        case _: assert_never(state)
```

I think we could use something analogous in this and other cases. What do you think?

Specifically for handling errors, maybe we could use the [Result](https://github.com/rustedpy/result) lib, that we've talked about before, now that we have `match` available. I also don't think exceptions would be very ergonomic.

Anyway, @msbrogli I'm fine leaving it like this, the logging is enough for now. But I suggest we keep this pattern in mind

Reply:
I did the coding to use typed errors with @glevco and here are the results:

1. It seems too much for this specific case.
2. We had a weird `mypy` error when attributing a variable inside the match.
3. The code seems to get more complex than it currently is.

Reply:
I'm OK with the current implementation.